dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt

```diff
@@ -0,0 +1,343 @@
+        }
+    }
+
+    val defaultColType = colTypes[ColType.DEFAULT]
+
+    // convert each ResultColumn to a DataColumn
+    val cols = csvReaderResult.map {
+        it.toDataColumn(
+            parserOptions = parserOptions,
+            desiredColType = colTypes[it.name()] ?: defaultColType,
+        )
+    }
+
+    return dataFrameOf(cols)
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun CsvReader.ResultColumn.toDataColumn(
+    parserOptions: ParserOptions,
+    desiredColType: ColType?,
+): DataColumn<*> {
+    val listSink = data()!! as ListSink
+    val columnData = listSink.data
+    val dataType = listSink.dataType
+    val hasNulls = listSink.hasNulls
+    val type = dataType().toKType().withNullability(hasNulls)
+
+    val column = DataColumn.createValueColumn(
+        name = name(),
+        values = columnData,
+        type = type,
+    )
+    if (dataType != STRING) return column
+
+    // attempt to perform additional parsing if necessary, will remain String if it fails
+    column as ValueColumn<String?>
+
+    val skipTypes = when {
+        desiredColType != null ->
+            // skip all types except the desired type
+            ParserOptions.allTypesExcept(desiredColType.toKType())
+
+        else ->
+            // no need to check for types that Deephaven already parses, skip those too
+            parserOptions.skipTypes + typesDeephavenAlreadyParses
+    }
+    val adjustsedParserOptions = parserOptions.copy(skipTypes = skipTypes)
+
+    return column.tryParse(adjustsedParserOptions)
+}
+
+private fun DataType?.toKType(): KType =
+    when (this) {
+        BOOLEAN_AS_BYTE -> typeOf<Boolean>()
+
+        // unused in Parsers.DEFAULT
+        BYTE -> typeOf<Byte>()
+
+        // unused in Parsers.DEFAULT
+        SHORT -> typeOf<Short>()
+
+        INT -> typeOf<Int>()
+
+        LONG -> typeOf<Long>()
+
+        // unused in Parsers.COMPLETE and Parsers.DEFAULT
+        FLOAT -> typeOf<Float>()
+
+        DOUBLE -> typeOf<Double>()
+
+        DATETIME_AS_LONG -> typeOf<LocalDateTime>()
+
+        CHAR -> typeOf<Char>()
+
+        STRING -> typeOf<String>()
+
+        // unused in Parsers.COMPLETE and Parsers.DEFAULT
+        TIMESTAMP_AS_LONG -> typeOf<LocalDateTime>()
+
+        DataType.CUSTOM -> error("custom data type")
+
+        null -> error("null data type")
+    }
+
+private fun legalizeHeader(header: Array<String>): Array<String> {
+    val generator = ColumnNameGenerator()
+    return header.map { generator.addUnique(it) }.toTypedArray()
+}
+
+private fun CsvSpecs.Builder.skipLines(takeHeaderFromCsv: Boolean, skipLines: Long): CsvSpecs.Builder =
+    if (takeHeaderFromCsv) {
+        skipHeaderRows(skipLines)
+    } else {
+        skipRows(skipLines)
+    }
+
+/**
+ * Sets the correct parsers for the csv, based on [colTypes] and [ParserOptions.skipTypes].
+ * If [ColType.DEFAULT] is present, it sets the default parser.
+ */
+private fun CsvSpecs.Builder.parsers(
```

Comment:
I also have one question, @Jolanrensen did you measure a JUnit test coverage here? When I am starting to think about multiple variety and combinatorial explosion of programming paths here, it makes me crazy.

Reply:
For this specific function, yes, all cases are covered.
If you narrow it down, there are fewer paths than you think:

- if no colTypes are given
  - let deephaven use all its default parsers
  - subtract parsers of skipTypes if those are supplied
- if colTypes are supplied
  - if DEFAULT is among the values
    - set the parser for each supplied column+coltype
    - let deephaven use _only_ the parser given as DEFAULT type
  - if DEFAULT is not among the values
    - set the parser for each supplied column+coltype
    - let deephaven use all its default parsers
    - subtract parsers of skipTypes if those are supplied

(you may notice that `skipTypes` will never skip a type explicitly set by `colTypes`. This is intended.)

I will add this overview to the kdoc too, It may be helpful when debugging.

I will add a failing date test though. (if it fails, it remains String)